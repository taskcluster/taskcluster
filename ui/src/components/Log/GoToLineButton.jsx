import React, { Component, Fragment } from 'react';
import { func } from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import NumericIcon from 'mdi-react/NumericIcon';
import TextField from '../TextField';
import Button from '../Button';

@withStyles(theme => ({
  fabIcon: {
    ...theme.mixins.fabIcon,
  },
}))
export default class GoToLineButton extends Component {
  static propTypes = {
    onLineNumberChange: func.isRequired,
  };

  state = {
    open: false,
    lineNumber: '',
  };

  handleChange = e => {
    const number = e.target.value;

    this.setState({
      isValid: number && !Number.isNaN(+number),
      lineNumber: number,
    });
  };

  handleClose = () => {
    this.setState({ open: false });
  };

  handleOpenClick = () => {
    this.setState({ open: true });
  };

  handleSubmit = e => {
    e.preventDefault();

    if (this.state.isValid) {
      this.setState({ open: false });
      this.props.onLineNumberChange(+this.state.lineNumber);
    }
  };

  render() {
    const { classes, onLineNumberChange, className, ...props } = this.props;
    const { open, isValid, lineNumber } = this.state;

    return (
      <Fragment>
        <Button
          size="small"
          spanProps={{ className }}
          tooltipProps={{ title: 'Go to Line' }}
          color="secondary"
          onClick={this.handleOpenClick}
          {...props}>
          <NumericIcon size={20} />
        </Button>
        <Dialog
          open={open}
          onClose={this.handleClose}
          aria-labelledby="go-to-line-title">
          <DialogTitle id="go-to-line-title">Go to line number</DialogTitle>
          <DialogContent>
            <form onSubmit={this.handleSubmit}>
              <TextField
                id="number"
                label="Line number"
                autoFocus
                fullWidth
                value={lineNumber}
                onChange={this.handleChange}
                type="number"
                margin="dense"
              />
            </form>
          </DialogContent>
          <DialogActions>
            <Button onClick={this.handleClose}>Cancel</Button>
            <Button
              onClick={this.handleSubmit}
              disabled={!isValid}
              variant="contained"
              color="secondary">
              Go to line
            </Button>
          </DialogActions>
        </Dialog>
      </Fragment>
    );
  }
}
