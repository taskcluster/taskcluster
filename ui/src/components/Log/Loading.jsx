import React, { Component } from 'react';
import Spinner from '@mozilla-frontend-infra/components/Spinner';
import { withStyles } from '@material-ui/core/styles';

@withStyles(theme => ({
  spinner: {
    marginTop: theme.spacing(3),
  },
}))
export default class Loading extends Component {
  render() {
    const { classes, ...props } = this.props;

    return <Spinner loading className={classes.spinner} {...props} />;
  }
}
