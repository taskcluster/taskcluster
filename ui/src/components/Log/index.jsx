import React, { Component, Fragment } from 'react';
import { arrayOf, bool, node, object, oneOfType, string } from 'prop-types';
import classNames from 'classnames';
import { withRouter } from 'react-router-dom';
import { LazyLog, ScrollFollow } from 'react-lazylog';
import storage from 'localforage';
import { omit } from 'ramda';
import { withStyles } from '@material-ui/core/styles';
import ArrowDownBoldCircleOutlineIcon from 'mdi-react/ArrowDownBoldCircleOutlineIcon';
import OpenInNewIcon from 'mdi-react/OpenInNewIcon';
import GoToLineButton from './GoToLineButton';
import Loading from './Loading';
import Button from '../Button';
import SpeedDial from '../SpeedDial';
import SpeedDialAction from '../SpeedDialAction';
import { THEME } from '../../utils/constants';

const LINE_NUMBER_MATCH = /L(\d+)-?(\d+)?/;
const FOLLOW_STORAGE_KEY = 'follow-log';

@withRouter
@withStyles(theme => {
  const filterStyles = {
    borderRadius: 4,
    padding: '4px 8px',
    lineHeight: 1,
    '&.active': {
      ...theme.mixins.successIcon,
      fill: THEME.PRIMARY_TEXT_DARK,
    },
  };

  return {
    '@global': {
      'div.react-lazylog': {
        backgroundColor: theme.palette.background.default,
        fontFamily: 'Consolas, Monaco, Andale Mono, Ubuntu Mono, monospace',
        fontSize: 13,
        paddingTop: 4,
        paddingBottom: theme.spacing(1),
        color: theme.palette.text.primary,
        '-webkit-font-smoothing': 'auto',
      },
      'div.react-lazylog-searchbar': {
        backgroundColor: theme.palette.background.default,
        fontFamily: 'Consolas, Monaco, Andale Mono, Ubuntu Mono, monospace',
        fontSize: 13,
        padding: 10,
        ...(theme.palette.type === 'light'
          ? {
              '& > .react-lazylog-searchbar-input': {
                height: theme.spacing(3),
                backgroundColor: theme.palette.grey['300'],
                color: theme.palette.common.black,
                borderColor: theme.palette.grey['300'],
              },
              '& > .react-lazylog-searchbar-filter': {
                ...filterStyles,
                '&.inactive': {
                  fill: theme.palette.text.disabled,
                },
              },
              '& > .react-lazylog-searchbar-matches': {
                '&.active': {
                  color: theme.palette.text.active,
                },
                '&.inactive': {
                  color: theme.palette.text.disabled,
                },
              },
            }
          : {
              '& > .react-lazylog-searchbar-input': {
                height: theme.spacing(3),
              },
              '& > .react-lazylog-searchbar-filter': filterStyles,
            }),
      },
    },
    highlight: {
      backgroundColor: `${theme.palette.action.selected} !important`,
    },
    line: {
      '& > a': {
        transition: theme.transitions.create('color'),
        color: theme.palette.text.inactive,
      },
      '&:hover > a': {
        color: `${theme.palette.text.primary} !important`,
      },
      '&$highlight > a': {
        color: theme.palette.text.primary,
      },
      '&:hover': {
        backgroundColor: `${theme.palette.action.hover} !important`,
      },
    },
    logActions: {
      position: 'absolute',
      top: 8,
      right: 380,
    },
    followButtonFollowing: {
      ...theme.mixins.successIcon,
    },
    followLogIconFollowing: {
      fill: THEME.PRIMARY_TEXT_DARK,
    },
    fabIcon: {
      ...theme.mixins.fabIcon,
    },
    logSpeedDial: {
      ...theme.mixins.fab,
      ...theme.mixins.actionButton,
      bottom: theme.spacing(3),
    },
    logToolbarButton: {
      width: 31,
      height: 25,
      minWidth: 'unset',
      padding: '0 6px',
    },
    goToLineButton: {
      marginRight: theme.spacing(1),
    },
  };
})
/**
 * Render a lazy-loading log viewer.
 */
export default class Log extends Component {
  static defaultProps = {
    stream: false,
    actions: null,
    GoToLineButtonProps: null,
    FollowLogButtonProps: null,
  };

  static propTypes = {
    /**
     * The remote URL for which to lazily render log text.
     */
    url: string.isRequired,
    /**
     * Specify `true` if the `url` is returning a partial progressive text
     * stream. Specify `false` if the `url` is a complete file.
     */
    stream: bool,
    /**
     * Render additional buttons in the actions area.
     */
    actions: oneOfType([node, arrayOf(node)]),
    /**
     * Specify props for the "Go to line" floating action button.
     */
    GoToLineButtonProps: object,
    /**
     * Specify props for the "Follow log" floating action button.
     */
    FollowLogButtonProps: object,
  };

  state = {
    lineNumber: null,
    follow: true,
  };

  getHighlightFromHash() {
    const hasHighlight = LINE_NUMBER_MATCH.exec(this.props.location.hash);
    const start = hasHighlight && +hasHighlight[1];
    const end = hasHighlight && +hasHighlight[2];

    return end ? [start, end] : start;
  }

  getScrollToLine() {
    if (typeof this.state.follow === 'boolean') {
      return null;
    }

    if (this.state.lineNumber) {
      return this.state.lineNumber;
    }

    const highlight = this.getHighlightFromHash();

    if (highlight) {
      return Array.isArray(highlight) ? highlight[0] : highlight;
    }
  }

  handleFollowClick = () => {
    const follow = !this.state.follow;

    storage.setItem(FOLLOW_STORAGE_KEY, follow);
    this.setState({ follow });
  };

  handleHighlight = range => {
    if (this.highlightRange && this.highlightRange.equals(range)) {
      return;
    }

    const first = range.first();
    const last = range.last();

    if (!first) {
      this.props.history.replace({ hash: '' });
    } else if (first === last) {
      this.props.history.replace({ hash: `#L${first}` });
    } else {
      this.props.history.replace({ hash: `#L${first}-${last}` });
    }

    this.highlightRange = range;
  };

  handleLineNumberChange = lineNumber => {
    this.setState({ lineNumber, follow: null });
  };

  handleScroll = ({ scrollTop, scrollHeight, clientHeight }) => {
    if (
      this.state.follow &&
      scrollHeight - scrollTop !== clientHeight &&
      // LazyLog triggers `handleScroll` on initial load.
      // This will make sure it doesn't set follow to false on log load.
      scrollTop !== 0
    ) {
      this.setState({ follow: false });
    }
  };

  shouldStartFollowing() {
    if (typeof this.state.follow === 'boolean') {
      return this.state.follow;
    }

    if (this.getScrollToLine()) {
      return false;
    }

    const pref = storage.getItem(FOLLOW_STORAGE_KEY);

    if (typeof pref === 'boolean') {
      return pref;
    }

    return false;
  }

  render() {
    const {
      url,
      stream,
      classes,
      actions,
      GoToLineButtonProps,
      FollowLogButtonProps,
      ...props
    } = this.props;
    const highlight = this.getHighlightFromHash();
    const scrollToLine = this.getScrollToLine();
    const containerStyle = {
      width: 'auto',
      maxWidth: 'initial',
      overflow: 'initial',
    };
    const rawLogButton = (
      <SpeedDialAction
        tooltipOpen
        icon={<OpenInNewIcon size={20} />}
        tooltipTitle="Raw Log"
        ButtonProps={{
          component: 'a',
          href: url,
          target: '_blank',
          rel: 'noopener noreferrer',
        }}
      />
    );
    const FollowLogButtonRest = omit(['className'], FollowLogButtonProps);

    return (
      <ScrollFollow
        startFollowing={this.shouldStartFollowing()}
        render={({ follow }) => (
          <Fragment>
            <LazyLog
              enableSearch
              caseInsensitive
              containerStyle={containerStyle}
              url={url}
              onScroll={this.handleScroll}
              stream={stream}
              selectableLines
              follow={follow}
              highlight={highlight}
              onHighlight={this.handleHighlight}
              scrollToLine={scrollToLine}
              scrollToAlignment="start"
              lineClassName={classes.line}
              highlightLineClassName={classes.highlight}
              loadingComponent={Loading}
              extraLines={5}
              {...props}
            />
            <div className={classes.logActions}>
              <GoToLineButton
                onLineNumberChange={this.handleLineNumberChange}
                className={classNames(
                  classes.logToolbarButton,
                  classes.goToLineButton
                )}
                {...GoToLineButtonProps}
              />
              <Button
                size="small"
                spanProps={{
                  className:
                    FollowLogButtonProps && FollowLogButtonProps.className,
                }}
                tooltipProps={{
                  title: follow && stream ? 'Unfollow Log' : 'Follow Log',
                }}
                className={classNames(classes.logToolbarButton, {
                  [classes.followButtonFollowing]: follow && stream,
                })}
                color={follow && stream ? 'inherit' : 'secondary'}
                onClick={this.handleFollowClick}
                {...FollowLogButtonRest}>
                <ArrowDownBoldCircleOutlineIcon
                  className={classNames({
                    [classes.followLogIconFollowing]: follow && stream,
                  })}
                />
              </Button>
            </div>
            {actions}
            <SpeedDial className={classes.logSpeedDial}>
              {rawLogButton}
            </SpeedDial>
          </Fragment>
        )}
      />
    );
  }
}
