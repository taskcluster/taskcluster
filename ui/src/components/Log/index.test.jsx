import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import Log from './index';

it('should render Log', () => {
  const { asFragment } = render(
    <MemoryRouter keyLength={0}>
      <Log url="https://localhost:1234/logs/public/test.log" />
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});

it('should render without following', () => {
  const { asFragment } = render(
    <MemoryRouter
      initialEntries={[
        '/tasks/fXIp54zURQaV1pMTsiTvzw/runs/0/logs/public/logs/live.log#L123',
      ]}>
      <Log url="https://localhost:1234/logs/public/test.log" />
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});
