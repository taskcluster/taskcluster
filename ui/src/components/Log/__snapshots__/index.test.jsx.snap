// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Log 1`] = `
<DocumentFragment>
  <div
    class="react-lazylog-searchbar _3hsz3SFOTsWryCiRA9Mjtq"
  >
    <input
      autocomplete="off"
      class="react-lazylog-searchbar-input _2JObJsQyVcVoB1YD6078ip"
      disabled=""
      name="search"
      placeholder="Search"
      type="text"
      value=""
    />
    <button
      class="react-lazylog-searchbar-filter inactive qZaVIaYTJWY-GaoXJJ2Cp _3cXa-36GYrmnrGIXEg46cX"
      disabled=""
    >
      <svg
        class="_1lq6olr9h2p7MJDCA8c-aL"
        viewBox="0 0 286.054 286.054"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.939 44.696h178.784a8.931 8.931 0 0 0 8.939-8.939V8.939A8.937 8.937 0 0 0 187.723 0H8.939C4.005 0 0 4.005 0 8.939v26.818c0 4.934 4.005 8.939 8.939 8.939zm268.176 35.757H8.939C4.005 80.453 0 84.457 0 89.392v26.818a8.937 8.937 0 0 0 8.939 8.939h268.176a8.931 8.931 0 0 0 8.939-8.939V89.392a8.936 8.936 0 0 0-8.939-8.939zM8.939 205.601h178.784a8.931 8.931 0 0 0 8.939-8.939v-26.818a8.931 8.931 0 0 0-8.939-8.939H8.939A8.937 8.937 0 0 0 0 169.844v26.818a8.937 8.937 0 0 0 8.939 8.939zm268.176 35.757H8.939A8.937 8.937 0 0 0 0 250.297v26.818a8.937 8.937 0 0 0 8.939 8.939h268.176a8.931 8.931 0 0 0 8.939-8.939v-26.818a8.931 8.931 0 0 0-8.939-8.939z"
        />
      </svg>
    </button>
    <span
      class="react-lazylog-searchbar-matches inactive _3cXa-36GYrmnrGIXEg46cX"
    >
      0 matches
    </span>
  </div>
  <div
    style="overflow: visible; height: 0px; width: 0px;"
  >
    <div
      aria-label="grid"
      aria-readonly="true"
      class="ReactVirtualized__Grid ReactVirtualized__List react-lazylog _3NRgTLvWbyfWD-1xileYWR"
      role="grid"
      style="box-sizing: border-box; direction: ltr; height: -45px; position: relative; width: 0px; will-change: transform; overflow-x: hidden; overflow-y: auto;"
      tabindex="0"
    />
  </div>
  <div
    class="Log-logActions-3"
  >
    <span
      class="Log-logToolbarButton-8 Log-goToLineButton-9"
      title="Go to Line"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-textSizeSmall MuiButton-sizeSmall"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <svg
            class="mdi-icon "
            fill="currentColor"
            height="20"
            viewBox="0 0 24 24"
            width="20"
          >
            <path
              d="M4,17V9H2V7H6V17H4M22,15C22,16.11 21.1,17 20,17H16V15H20V13H18V11H20V9H16V7H20A2,2 0 0,1 22,9V10.5A1.5,1.5 0 0,1 20.5,12A1.5,1.5 0 0,1 22,13.5V15M14,15V17H8V13C8,11.89 8.9,11 10,11H12V9H8V7H12A2,2 0 0,1 14,9V11C14,12.11 13.1,13 12,13H10V15H14Z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </span>
    <span
      class=""
      title="Follow Log"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text Log-logToolbarButton-8 Log-followButtonFollowing-4 MuiButton-textSecondary MuiButton-textSizeSmall MuiButton-sizeSmall"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <svg
            class="mdi-icon Log-followLogIconFollowing-5"
            fill="currentColor"
            height="24"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M12,17L7,12H10V8H14V12H17L12,17M12,2A10,10 0 0,1 22,12A10,10 0 0,1 12,22A10,10 0 0,1 2,12A10,10 0 0,1 12,2M12,4A8,8 0 0,0 4,12A8,8 0 0,0 12,20A8,8 0 0,0 20,12A8,8 0 0,0 12,4Z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </span>
  </div>
  <div
    class="MuiSpeedDial-root MuiSpeedDial-directionUp SpeedDial-speedDial-11 Log-logSpeedDial-7"
    role="presentation"
  >
    <button
      aria-controls="speed-dial-actions"
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="speed-dial"
      class="MuiButtonBase-root MuiFab-root MuiSpeedDial-fab MuiFab-secondary"
      style="transform: none; webkit-transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiFab-label"
      >
        <span
          class="MuiSpeedDialIcon-root"
        >
          <svg
            class="mdi-icon MuiSpeedDialIcon-openIcon"
            fill="currentColor"
            height="24"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"
            />
          </svg>
          <svg
            class="mdi-icon MuiSpeedDialIcon-icon"
            fill="currentColor"
            height="24"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M12,16A2,2 0 0,1 14,18A2,2 0 0,1 12,20A2,2 0 0,1 10,18A2,2 0 0,1 12,16M12,10A2,2 0 0,1 14,12A2,2 0 0,1 12,14A2,2 0 0,1 10,12A2,2 0 0,1 12,10M12,4A2,2 0 0,1 14,6A2,2 0 0,1 12,8A2,2 0 0,1 10,6A2,2 0 0,1 12,4Z"
            />
          </svg>
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div
      aria-orientation="vertical"
      class="MuiSpeedDial-actions MuiSpeedDial-actionsClosed"
      id="speed-dial-actions"
      role="menu"
    >
      <span
        class="MuiSpeedDialAction-staticTooltip MuiSpeedDialAction-tooltipPlacementLeft MuiSpeedDialAction-staticTooltipClosed"
        id="speed-dial-action-0"
      >
        <span
          class="MuiSpeedDialAction-staticTooltipLabel SpeedDialAction-staticTooltipLabel-14"
          id="speed-dial-action-0-label"
          style="transition-delay: 30ms;"
        >
          Raw Log
        </span>
        <a
          aria-describedby="speed-dial-action-0-label"
          aria-disabled="false"
          class="MuiButtonBase-root MuiFab-root MuiSpeedDialAction-fab SpeedDialAction-icon-12 MuiSpeedDialAction-fabClosed MuiFab-sizeSmall"
          href="https://localhost:1234/logs/public/test.log"
          rel="noopener noreferrer"
          role="menuitem"
          style="transition-delay: 30ms;"
          tabindex="-1"
          target="_blank"
        >
          <span
            class="MuiFab-label"
          >
            <svg
              class="mdi-icon "
              fill="currentColor"
              height="20"
              viewBox="0 0 24 24"
              width="20"
            >
              <path
                d="M14,3V5H17.59L7.76,14.83L9.17,16.24L19,6.41V10H21V3M19,19H5V5H12V3H5C3.89,3 3,3.9 3,5V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V12H19V19Z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </a>
      </span>
    </div>
  </div>
  <div
    class="resize-triggers"
  >
    <div
      class="expand-trigger"
    >
      <div
        style="width: 1px; height: 1px;"
      />
    </div>
    <div
      class="contract-trigger"
    />
  </div>
</DocumentFragment>
`;

exports[`should render without following 1`] = `
<DocumentFragment>
  <div
    class="react-lazylog-searchbar _3hsz3SFOTsWryCiRA9Mjtq"
  >
    <input
      autocomplete="off"
      class="react-lazylog-searchbar-input _2JObJsQyVcVoB1YD6078ip"
      disabled=""
      name="search"
      placeholder="Search"
      type="text"
      value=""
    />
    <button
      class="react-lazylog-searchbar-filter inactive qZaVIaYTJWY-GaoXJJ2Cp _3cXa-36GYrmnrGIXEg46cX"
      disabled=""
    >
      <svg
        class="_1lq6olr9h2p7MJDCA8c-aL"
        viewBox="0 0 286.054 286.054"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.939 44.696h178.784a8.931 8.931 0 0 0 8.939-8.939V8.939A8.937 8.937 0 0 0 187.723 0H8.939C4.005 0 0 4.005 0 8.939v26.818c0 4.934 4.005 8.939 8.939 8.939zm268.176 35.757H8.939C4.005 80.453 0 84.457 0 89.392v26.818a8.937 8.937 0 0 0 8.939 8.939h268.176a8.931 8.931 0 0 0 8.939-8.939V89.392a8.936 8.936 0 0 0-8.939-8.939zM8.939 205.601h178.784a8.931 8.931 0 0 0 8.939-8.939v-26.818a8.931 8.931 0 0 0-8.939-8.939H8.939A8.937 8.937 0 0 0 0 169.844v26.818a8.937 8.937 0 0 0 8.939 8.939zm268.176 35.757H8.939A8.937 8.937 0 0 0 0 250.297v26.818a8.937 8.937 0 0 0 8.939 8.939h268.176a8.931 8.931 0 0 0 8.939-8.939v-26.818a8.931 8.931 0 0 0-8.939-8.939z"
        />
      </svg>
    </button>
    <span
      class="react-lazylog-searchbar-matches inactive _3cXa-36GYrmnrGIXEg46cX"
    >
      0 matches
    </span>
  </div>
  <div
    style="overflow: visible; height: 0px; width: 0px;"
  >
    <div
      aria-label="grid"
      aria-readonly="true"
      class="ReactVirtualized__Grid ReactVirtualized__List react-lazylog _3NRgTLvWbyfWD-1xileYWR"
      role="grid"
      style="box-sizing: border-box; direction: ltr; height: -45px; position: relative; width: 0px; will-change: transform; overflow-x: hidden; overflow-y: auto;"
      tabindex="0"
    />
  </div>
  <div
    class="Log-logActions-17"
  >
    <span
      class="Log-logToolbarButton-22 Log-goToLineButton-23"
      title="Go to Line"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-textSizeSmall MuiButton-sizeSmall"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <svg
            class="mdi-icon "
            fill="currentColor"
            height="20"
            viewBox="0 0 24 24"
            width="20"
          >
            <path
              d="M4,17V9H2V7H6V17H4M22,15C22,16.11 21.1,17 20,17H16V15H20V13H18V11H20V9H16V7H20A2,2 0 0,1 22,9V10.5A1.5,1.5 0 0,1 20.5,12A1.5,1.5 0 0,1 22,13.5V15M14,15V17H8V13C8,11.89 8.9,11 10,11H12V9H8V7H12A2,2 0 0,1 14,9V11C14,12.11 13.1,13 12,13H10V15H14Z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </span>
    <span
      class=""
      title="Follow Log"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text Log-logToolbarButton-22 MuiButton-textSecondary MuiButton-textSizeSmall MuiButton-sizeSmall"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          <svg
            class="mdi-icon "
            fill="currentColor"
            height="24"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M12,17L7,12H10V8H14V12H17L12,17M12,2A10,10 0 0,1 22,12A10,10 0 0,1 12,22A10,10 0 0,1 2,12A10,10 0 0,1 12,2M12,4A8,8 0 0,0 4,12A8,8 0 0,0 12,20A8,8 0 0,0 20,12A8,8 0 0,0 12,4Z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </span>
  </div>
  <div
    class="MuiSpeedDial-root MuiSpeedDial-directionUp SpeedDial-speedDial-25 Log-logSpeedDial-21"
    role="presentation"
  >
    <button
      aria-controls="speed-dial-actions"
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="speed-dial"
      class="MuiButtonBase-root MuiFab-root MuiSpeedDial-fab MuiFab-secondary"
      style="transform: none; webkit-transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: transform 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiFab-label"
      >
        <span
          class="MuiSpeedDialIcon-root"
        >
          <svg
            class="mdi-icon MuiSpeedDialIcon-openIcon"
            fill="currentColor"
            height="24"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"
            />
          </svg>
          <svg
            class="mdi-icon MuiSpeedDialIcon-icon"
            fill="currentColor"
            height="24"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M12,16A2,2 0 0,1 14,18A2,2 0 0,1 12,20A2,2 0 0,1 10,18A2,2 0 0,1 12,16M12,10A2,2 0 0,1 14,12A2,2 0 0,1 12,14A2,2 0 0,1 10,12A2,2 0 0,1 12,10M12,4A2,2 0 0,1 14,6A2,2 0 0,1 12,8A2,2 0 0,1 10,6A2,2 0 0,1 12,4Z"
            />
          </svg>
        </span>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div
      aria-orientation="vertical"
      class="MuiSpeedDial-actions MuiSpeedDial-actionsClosed"
      id="speed-dial-actions"
      role="menu"
    >
      <span
        class="MuiSpeedDialAction-staticTooltip MuiSpeedDialAction-tooltipPlacementLeft MuiSpeedDialAction-staticTooltipClosed"
        id="speed-dial-action-0"
      >
        <span
          class="MuiSpeedDialAction-staticTooltipLabel SpeedDialAction-staticTooltipLabel-28"
          id="speed-dial-action-0-label"
          style="transition-delay: 30ms;"
        >
          Raw Log
        </span>
        <a
          aria-describedby="speed-dial-action-0-label"
          aria-disabled="false"
          class="MuiButtonBase-root MuiFab-root MuiSpeedDialAction-fab SpeedDialAction-icon-26 MuiSpeedDialAction-fabClosed MuiFab-sizeSmall"
          href="https://localhost:1234/logs/public/test.log"
          rel="noopener noreferrer"
          role="menuitem"
          style="transition-delay: 30ms;"
          tabindex="-1"
          target="_blank"
        >
          <span
            class="MuiFab-label"
          >
            <svg
              class="mdi-icon "
              fill="currentColor"
              height="20"
              viewBox="0 0 24 24"
              width="20"
            >
              <path
                d="M14,3V5H17.59L7.76,14.83L9.17,16.24L19,6.41V10H21V3M19,19H5V5H12V3H5C3.89,3 3,3.9 3,5V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V12H19V19Z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </a>
      </span>
    </div>
  </div>
  <div
    class="resize-triggers"
  >
    <div
      class="expand-trigger"
    >
      <div
        style="width: 1px; height: 1px;"
      />
    </div>
    <div
      class="contract-trigger"
    />
  </div>
</DocumentFragment>
`;
